### PR TITLE
Fix client not being dropped on shutdown

### LINKED ISSUE #6213: pure virtual method called + terminate called recursively
Exiting parity version Parity/v1.7.0-beta-5f2cabd-20170727/arm-linux-gnu/rustc1.18.0 results in the following messages. 

> 2017-08-02 10:42:37 UTC Finishing work, please wait...
> pure virtual method called
> pure virtual method called
> pure virtual method called
> terminate called recursively

Running using:

> parity -d /home/.../data/Ethereum --pruning archive --public-node

**Logs**
Here is the tail end of the logs in archive/db

> 2017/08/02-10:40:56.165605 6c3ff020 (Original Log Time 2017/08/02-10:40:56.165478) [col0] Level summary: files[9 0 0 0 0 0 42] max score 0.50
> 2017/08/02-10:41:28.990260 71dff020 Flushing all column families with data in WAL number 41696. Total log size is 67170274 while max_total_wal_size is 67108864
> 2017/08/02-10:41:29.140767 71dff020 [col6] New memtable created with log file: #41913. Immutable memtables: 0.
> 2017/08/02-10:41:29.140931 6cbff020 [JOB 6357] Syncing log #41911
> 2017/08/02-10:41:29.560927 6cbff020 (Original Log Time 2017/08/02-10:41:29.140901) Calling FlushMemTableToOutputFile with column family [col6], flush slots available 2, compaction slots allowed 1, compaction slots scheduled 0
> 2017/08/02-10:41:29.560960 6cbff020 [col6] [JOB 6357] Flushing memtable with next log file: 41913
> 2017/08/02-10:41:29.561055 6cbff020 EVENT_LOG_v1 {"time_micros": 1501670489561011, "job": 6357, "event": "flush_started", "num_memtables": 1, "num_entries": 2, "num_deletes": 0, "memory_usage": 2208}
> 2017/08/02-10:41:29.561067 6cbff020 [col6] [JOB 6357] Level-0 flush table #41914: started
> 2017/08/02-10:41:29.773092 6cbff020 EVENT_LOG_v1 {"time_micros": 1501670489772948, "cf_name": "col6", "job": 6357, "event": "table_file_creation", "file_number": 41914, "file_size": 879, "table_properties": {"data_size": 35, "index_size": 27, "filter_size": 0, "raw_key_size": 17, "raw_average_key_size": 17, "raw_value_size": 2, "raw_average_value_size": 2, "num_data_blocks": 1, "num_entries": 1, "filter_policy_name": "", "kDeletedKeys": "0", "kMergeOperands": "0"}}
> 2017/08/02-10:41:29.801913 6cbff020 [col6] [JOB 6357] Level-0 flush table #41914: 879 bytes OK
> 2017/08/02-10:41:30.082633 6cbff020 (Original Log Time 2017/08/02-10:41:29.909607) [col6] Level-0 commit table #41914 started
> 2017/08/02-10:41:30.082648 6cbff020 (Original Log Time 2017/08/02-10:41:30.082349) [col6] Level-0 commit table #41914: memtable #1 done
> 2017/08/02-10:41:30.082658 6cbff020 (Original Log Time 2017/08/02-10:41:30.082448) EVENT_LOG_v1 {"time_micros": 1501670490082407, "job": 6357, "event": "flush_finished", "lsm_state": [2, 0, 0, 0, 0, 0, 1], "immutable_memtables": 0}
> 2017/08/02-10:41:30.082668 6cbff020 (Original Log Time 2017/08/02-10:41:30.082523) [col6] Level summary: files[2 0 0 0 0 0 1] max score 0.75
> 2017/08/02-10:41:30.082939 6cbff020 [JOB 6357] Try to delete WAL files size 43710716, prev total WAL file size 67194125, number of live WAL files 88.
> 2017/08/02-10:41:38.705955 72dff020 Flushing column family with largest mem table size. Write buffer is using 16211688138490200064 bytes out of a total of 18184929890698486689.
> 2017/08/02-10:41:38.865260 72dff020 [col0] New memtable created with log file: #41915. Immutable memtables: 0.
> 2017/08/02-10:41:38.865361 72dff020 [WARN] [col0] Increasing compaction threads because we have 10 level-0 files
> 2017/08/02-10:41:38.865517 6c3ff020 [JOB 6358] Syncing log #41913
> 2017/08/02-10:41:39.190184 6c3ff020 (Original Log Time 2017/08/02-10:41:38.865490) Calling FlushMemTableToOutputFile with column family [col0], flush slots available 2, compaction slots allowed 1, compaction slots scheduled 0
> 2017/08/02-10:41:39.190214 6c3ff020 [col0] [JOB 6358] Flushing memtable with next log file: 41915
> 2017/08/02-10:41:39.190306 6c3ff020 EVENT_LOG_v1 {"time_micros": 1501670499190263, "job": 6358, "event": "flush_started", "num_memtables": 1, "num_entries": 1629, "num_deletes": 0, "memory_usage": 642400}
> 2017/08/02-10:41:39.190317 6c3ff020 [col0] [JOB 6358] Level-0 flush table #41916: started
> 2017/08/02-10:41:43.027820 6c3ff020 EVENT_LOG_v1 {"time_micros": 1501670503027658, "cf_name": "col0", "job": 6358, "event": "table_file_creation", "file_number": 41916, "file_size": 623726, "table_properties": {"data_size": 620417, "index_size": 3453, "filter_size": 0, "raw_key_size": 64540, "raw_average_key_size": 39, "raw_value_size": 559779, "raw_average_value_size": 346, "num_data_blocks": 158, "num_entries": 1614, "filter_policy_name": "", "kDeletedKeys": "0", "kMergeOperands": "0"}}
> 2017/08/02-10:41:43.048227 6c3ff020 [col0] [JOB 6358] Level-0 flush table #41916: 623726 bytes OK
> 2017/08/02-10:41:43.349171 6c3ff020 [WARN] [col0] Increasing compaction threads because we have 11 level-0 files
> 2017/08/02-10:41:43.349316 6c3ff020 (Original Log Time 2017/08/02-10:41:43.187860) [col0] Level-0 commit table #41916 started
> 2017/08/02-10:41:43.349327 6c3ff020 (Original Log Time 2017/08/02-10:41:43.349060) [col0] Level-0 commit table #41916: memtable #1 done
> 2017/08/02-10:41:43.349337 6c3ff020 (Original Log Time 2017/08/02-10:41:43.349146) EVENT_LOG_v1 {"time_micros": 1501670503349108, "job": 6358, "event": "flush_finished", "lsm_state": [10, 0, 0, 0, 0, 0, 42], "immutable_memtables": 0}
> 2017/08/02-10:41:43.349347 6c3ff020 (Original Log Time 2017/08/02-10:41:43.349223) [col0] Level summary: files[10 0 0 0 0 0 42] max score 0.75
> 2017/08/02-10:42:19.902672 70dff020 Flushing column family with largest mem table size. Write buffer is using 4297015296 bytes out of a total of 5781852724.
> 2017/08/02-10:42:20.127455 70dff020 [col2] New memtable created with log file: #41917. Immutable memtables: 0.
> 2017/08/02-10:42:20.127796 6cbff020 [JOB 6359] Syncing log #41915
> 2017/08/02-10:42:20.516100 6cbff020 (Original Log Time 2017/08/02-10:42:20.127766) Calling FlushMemTableToOutputFile with column family [col2], flush slots available 2, compaction slots allowed 1, compaction slots scheduled 0
> 2017/08/02-10:42:20.516151 6cbff020 [col2] [JOB 6359] Flushing memtable with next log file: 41917
> 2017/08/02-10:42:20.516305 6cbff020 EVENT_LOG_v1 {"time_micros": 1501670540516226, "job": 6359, "event": "flush_started", "num_memtables": 1, "num_entries": 384, "num_deletes": 0, "memory_usage": 623740}
> 2017/08/02-10:42:20.516327 6cbff020 [col2] [JOB 6359] Level-0 flush table #41918: started
> 2017/08/02-10:42:27.181659 6cbff020 EVENT_LOG_v1 {"time_micros": 1501670547181500, "cf_name": "col2", "job": 6359, "event": "table_file_creation", "file_number": 41918, "file_size": 431444, "table_properties": {"data_size": 429000, "index_size": 2185, "filter_size": 0, "raw_key_size": 15360, "raw_average_key_size": 40, "raw_value_size": 602384, "raw_average_value_size": 1568, "num_data_blocks": 101, "num_entries": 384, "filter_policy_name": "", "kDeletedKeys": "0", "kMergeOperands": "0"}}
> 2017/08/02-10:42:27.229476 6cbff020 [col2] [JOB 6359] Level-0 flush table #41918: 431444 bytes OK
> 2017/08/02-10:42:27.600216 6cbff020 (Original Log Time 2017/08/02-10:42:27.345107) [col2] Level-0 commit table #41918 started
> 2017/08/02-10:42:27.600244 6cbff020 (Original Log Time 2017/08/02-10:42:27.599825) [col2] Level-0 commit table #41918: memtable #1 done
> 2017/08/02-10:42:27.600264 6cbff020 (Original Log Time 2017/08/02-10:42:27.599972) EVENT_LOG_v1 {"time_micros": 1501670547599902, "job": 6359, "event": "flush_finished", "lsm_state": [1, 0, 1, 1, 1, 1, 1], "immutable_memtables": 0}
> 2017/08/02-10:42:27.600283 6cbff020 (Original Log Time 2017/08/02-10:42:27.600066) [col2] Level summary: files[1 0 1 1 1 1 1] max score 1.50

Let me know if you need some other logs.

### PR DESCRIPTION
**Pull request by andresilva**
I noticed that the `Client` was not being dropped on shutdown, this also meant that the `Database` was not dropped and that RocksDB wasn't properly closed. I'm not sure why it's necessary to wait for the `Client` to drop, maybe there's some other thread holding a reference to it that takes a while to die?

I started by converting some stuff to `Weak` on the RPC clients and it fixed the issue. And then I implemented the wait mechanism (suggested by @tomusdrw) and I noticed that declaring `Weak` references was no longer necessary.

It took me a really long time to figure this out (and incremental compilation really helped). 😐 

Should fix #6213, #7518, #7334.

### CODE REVIEW

**Comment by andresilva** [comment]
Please test this locally by adding a `println` to the `Database` `Drop` instance. Also check if the timeouts for shutdown are reasonable.

File: parity/run.rs [hunk 1]
```diff
@@ -856,6 +848,9 @@ pub fn execute(cmd: RunCmd, can_restart: bool, logger: Arc<RotatingLogger>) -> R
 		open_dapp(&cmd.dapps_conf, &cmd.http_conf, &dapp)?;
 	}
 
+	// Create a weak reference to the client so that we can wait on shutdown until it is dropped
```
Comment (tomusdrw) [inline parity/run.rs]: Something wierd with the indentation

File: parity/run.rs [hunk 2]
```diff
@@ -869,7 +864,34 @@ pub fn execute(cmd: RunCmd, can_restart: bool, logger: Arc<RotatingLogger>) -> R
...
+		if !TcpListener::bind(&addr as &str).is_ok() {
+			return open_ui(&cmd.ws_conf, &cmd.ui_conf, &cmd.logger_config).map(|_| (false, None));
+		}
+	}
+
+	// increase max number of open files
+	raise_fd_limit();
+
+	fn wait<T>(res: Result<((bool, Option<String>), Weak<T>), String>) -> Result<(bool, Option<String>), String> {
+		res.map(|r| {
```
Comment (tomusdrw) [inline parity/run.rs]: You should be able to destructure here:
```
|(restart, weak_client)| {
  ...
}
```

File: parity/run.rs [hunk 3]
```diff
@@ -1002,3 +1024,32 @@ fn wait_for_exit(
 	let _ = exit.1.wait(&mut l);
 	l.clone()
 }
+
+fn wait_for_drop<T>(w: Weak<T>) {
+	let sleep_duration = Duration::from_secs(1);
+	let warn_timeout = Duration::from_secs(30);
+	let max_timeout = Duration::from_secs(120);
```
Comment (tomusdrw) [inline parity/run.rs]: IMHO that's a bit too low (especially on HDD). Can we have like 5minutes here?

File: parity/run.rs [hunk 4]
```diff
@@ -1002,3 +1024,32 @@ fn wait_for_exit(
...
+
+	let instant = Instant::now();
+	let mut warned = false;
+
+	loop {
+		{
```
Comment (tomusdrw) [inline parity/run.rs]: Why do we need that additional block?

File: parity/run.rs [hunk 5]
```diff
@@ -1002,3 +1024,32 @@ fn wait_for_exit(
...
+			if let None = w.upgrade() {
+				break;
+			}
+		}
+
+		if instant.elapsed() > max_timeout {
```
Comment (tomusdrw) [inline parity/run.rs]: Nitpick. Imho this loop is more concise and cleaner for me (but don't have strong opinion):
```rust
while instance.elapsed() < max_timeout {
   if w.upgrade().is_none() {
     return;
   }
   if !warned && instant.elapsed() > warn_timeout {
     warned = true;
     warn!(...);
   }
}
warn('unclean');
```

File: parity/run.rs (hunk 4 above)
Comment (andresilva) [reply parity/run.rs]: If the `upgrade` is successful is it dropped right away or only at the end of the loop iteration? I just wanted to make sure it was dropped right away (before sleeping).

File: parity/run.rs (hunk 1 above)
Comment (andresilva) [reply parity/run.rs]: Spaces.

**Comment by svyatonik** [comment]
My two cents: tested it for some time (by adding `println`) yesterday - have seen the printed line every time. Haven't tested with light client, though.

**Pull request merged**